scylla/src/transport/session.rs

```diff
@@ -1814,13 +1807,63 @@ impl Session {
         prepared: &PreparedStatement,
         serialized_values: &SerializedValues,
     ) -> Result<Option<Token>, QueryError> {
-        match self.calculate_partition_key(prepared, serialized_values) {
-            Ok(Some(partition_key)) => {
-                let partitioner_name = prepared.get_partitioner_name();
-                Ok(Some(partitioner_name.hash(&partition_key)))
-            }
-            Ok(None) => Ok(None),
-            Err(err) => Err(err),
+        Ok(self
+            .calculate_partition_key(prepared, serialized_values)?
+            .map(|partition_key| prepared.get_partitioner_name().hash(&partition_key)))
+    }
+
+    /// Get the first node/shard that the load balancer would target if running this query
+    ///
```

Comment:
Although the shard is fixed for a given token (at least until resharding occurs), the first node returned by a load balancer may, and likely will, vary from call to call. Taking a look at the default load balancer, `DefaultPolicy`, we can see that we choose a random replica for `pick()`, so `first_shard_for_statement()` may return a replica different that one that will be queried later, while executing the batch of interest.

Reply:
Ah, that seems to be a problem indeed for [this approach](https://github.com/scylladb/scylla-rust-driver/issues/448#issuecomment-1125008817).
It seems that the only thing left for it to work ideally and provide the benefits of immensely reducing network-management CPU-cost while preserving shard awareness would be to make it deterministic (at least on the second run).

The main option I see for this would be to use some other part of the token to choose among the replicas.

Reply:
It looks like another option would be to enforce the target shard the second time around by setting a custom `LoadBalancerPolicy` in `Batch.config.execution_profile_handle.load_balancing_policy`.

Reply:
I've implemented that second option.